system/lib/wasmfs/file_table.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include <assert.h>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include <wasi/api.h>
+
+namespace wasmfs {
+
+class FileDescription {
+  std::shared_ptr<File> file;
+  __wasi_filedelta_t offset;
+  std::mutex mutex;
+
+public:
+  FileDescription(uint32_t offset, std::shared_ptr<File> file) : offset(offset), file(file) {}
+
+  class Handle {
+    FileDescription& fileDescription;
+    std::unique_lock<std::mutex> lock;
+
+  public:
+    Handle(FileDescription& fileDescription)
+      : fileDescription(fileDescription), lock(fileDescription.mutex) {}
+
+    std::shared_ptr<File>& getFile();
+  };
+
+  Handle get() { return Handle(*this); }
+};
+
+class FileTable {
+  static std::vector<std::shared_ptr<FileDescription>> entries;
+  std::mutex mutex;
+
+  FileTable();
+
+public:
+  // Handle represents an RAII wrapper object. Access to the global FileTable must go through a
+  // Handle. A Handle holds the single global FileTable's lock for the duration of its lifetime.
+  // This is necessary because a FileTable may have atomic oeprations where the lock must be held
+  // across multiple methods. By providing access through the handle, callers of file table methods
+  // do not need to remember to take a lock for every access.
+  class Handle {
+    FileTable& fileTable;
+    std::unique_lock<std::mutex> lock;
+
+  public:
+    Handle(FileTable& fileTable) : fileTable(fileTable), lock(fileTable.mutex) {}
+
+    // Adds given FileDescription to FileTable entries. Returns fd (insertion index in entries).
+    __wasi_fd_t add(std::shared_ptr<FileDescription> ptr);
+
+    // Removes FileDescription in FileTable entries corresponding to given fd.
+    void remove(__wasi_fd_t fd);
+
+    // Entry is used to override the subscript [] operator.
+    // This allows the user to get and set values in the FileTable entries vector.
```

Comment:
I had a question about this before that I can't seem to find now - got lost in the edits I guess. I asked about "out of bounds" there, and I see the text has been edited, but I still don't get this - why can't the user get and set values in the file table entries vector in a more straightforward way - directly, or using a simple API? An entire class here seems odd.